### PR TITLE
track billing info stats

### PR DESCRIPTION
**Pull request by bemaxima**
https://github.com/whynotearth/meredith-core/issues/183

Note: the webhook route has been changed.

### CODE REVIEW

File: WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs [hunk 1]
```diff
@@ -38,6 +40,29 @@ public async Task<OkResult> Create(List<EventItem> events)
...
+            {
+                return NotFound($"Company '{companySlug}' not found.");
+            }
+            var lastMonth = DateTime.Now.AddMonths(-1);
+            var monthlySentEmails = await _dbContext.EmailRecipients.Where(item =>
+                item.CompanyId == company.Id).CountAsync();
+
+            var monthlyActiveUsers = await _dbContext.EmailRecipients
+                .Where(item =>
+                    item.Status == EmailStatus.Opened &&
```
Comment (ctyar) [inline WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs:57]:
```suggestion
                    item.Status >= EmailStatus.Opened &&
```

File: WhyNotEarth.Meredith.App/Results/Api/v0/Public/SendGrid/StatResult.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+﻿namespace WhyNotEarth.Meredith.App.Results.Api.v0.Public.SendGrid
+{
+    public class StatResult
```
Comment (ctyar) [inline WhyNotEarth.Meredith.App/Results/Api/v0/Public/SendGrid/StatResult.cs:3]: The name is too general, maybe EmailStatsResult

File: WhyNotEarth.Meredith.App/Results/Api/v0/Public/SendGrid/StatResult.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    {
+        public int MonthlyActiveUsers { get; set; }
+        public int MonthlySentEmails { get; set; }
```
Comment (ctyar) [inline WhyNotEarth.Meredith.App/Results/Api/v0/Public/SendGrid/StatResult.cs:6]: I like to make the properties get only and set them all in the constructor so we wont miss any of them by accident

File: WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs [hunk 4]
```diff
@@ -1,18 +1,19 @@
...
 using WhyNotEarth.Meredith.Data.Entity;
-using WhyNotEarth.Meredith.Data.Entity.Models.Modules.Volkswagen;
+using WhyNotEarth.Meredith.Data.Entity.Models;
 
-namespace WhyNotEarth.Meredith.App.Controllers.Api.v0.Volkswagen
+namespace WhyNotEarth.Meredith.App.Controllers.Api.v0.Public
 {
     [ApiVersion("0")]
-    [Route("api/v0/volkswagen/memo/sendgrid/webhook")]
+    [Route("api/v0/sendgrid")]
```
Comment (ctyar) [inline WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs:15]: I think we should rename the controller/route to something more general, like `emails` and `emails/stats` for the action

File: WhyNotEarth.Meredith.Data.Entity/Migrations/20200520085040_AddCompanyIdToEmailReceipients.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            migrationBuilder.AddColumn<int>(
+                name: "CompanyId",
+                schema: "ModuleVolkswagen",
+                table: "EmailRecipients",
+                nullable: false,
+                defaultValue: 0);
+
+            migrationBuilder.Sql(
+                "UPDATE \"ModuleVolkswagen\".\"EmailRecipients\"" +
+                "SET \"CompanyId\" = (SELECT \"Id\" FROM public.\"Companies\" WHERE \"Slug\" = 'Volkswagen' limit 1)"
```
Comment (ctyar) [inline WhyNotEarth.Meredith.Data.Entity/Migrations/20200520085040_AddCompanyIdToEmailReceipients.cs:18]: `Volkswagen` should be lower case

File: WhyNotEarth.Meredith/DependencyInjection/ServiceProviderExtensions.cs [hunk 6]
```diff
@@ -27,7 +27,8 @@ public static IServiceCollection AddMeredith(this IServiceCollection serviceColl
                 .AddScoped<SendGridService>()
                 .AddScoped<IUserManager, UserManager>()
                 .AddScoped<GoogleStorageService>()
-                .AddScoped<SettingsService>();
+                .AddScoped<SettingsService>()
+                .AddScoped<ProductService>();
```
Comment (ctyar) [inline WhyNotEarth.Meredith/DependencyInjection/ServiceProviderExtensions.cs:31]: Is this necessary?

File: WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs (hunk 4 above)
Comment (bemaxima) [reply WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs:15]: Do you think that 'SendGrid' is necessary for the route?
Do these routes work for you?

- "api/v0/emails/webhook"
- "api/v0/emails/stats"

File: WhyNotEarth.Meredith/DependencyInjection/ServiceProviderExtensions.cs (hunk 6 above)
Comment (bemaxima) [reply WhyNotEarth.Meredith/DependencyInjection/ServiceProviderExtensions.cs:31]: I have no idea when the line has been added :|

File: WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs [hunk 7]
```diff
@@ -38,6 +40,29 @@ public async Task<OkResult> Create(List<EventItem> events)
...
+        public async Task<ActionResult<StatResult>> EmailStats(string companySlug)
+        {
+            var company = await _dbContext.Companies.FirstOrDefaultAsync(item => item.Name == companySlug);
+            if(company is null)
+            {
+                return NotFound($"Company '{companySlug}' not found.");
+            }
+            var lastMonth = DateTime.Now.AddMonths(-1);
+            var monthlySentEmails = await _dbContext.EmailRecipients.Where(item =>
+                item.CompanyId == company.Id).CountAsync();
```
[the remark this answers is not in the document]
Comment (bemaxima) [reply WhyNotEarth.Meredith.App/Controllers/Api/v0/Public/SendGridController.cs:53]: Something isn't right. I saw this issue before the commit and fix it.

**Comment by ctyar** [comment]
Thank you @bemaxima 
Merged in https://github.com/whynotearth/meredith-core/commit/a370620481183fd5aec8a2a8f538eee6adb24f87

**Pull request closed without merging**